libass/ass_render.c

```diff
@@ -2413,6 +2424,10 @@ static void render_and_combine_glyphs(ASS_Renderer *render_priv,
         if (info->effect_type == EF_KARAOKE_KF)
             info->effect_timing = lround(d6_to_double(info->leftmost_x) +
                 d6_to_double(info->effect_timing) * render_priv->font_scale_x);
+        else if (info->effect_type == EF_NONE || info->effect_timing > 0)
+            info->filter.alpha_secondary = info->filter.alpha_primary;
+        else
+            info->filter.alpha_primary = info->filter.alpha_secondary;
```

Comment:
For \k and \ko, I’m setting `alpha_primary` and `alpha_secondary` to the same number in the cache key. This means that the bitmap will be re-rendered when the karaoke time arrives (which happens at most once), but on the other hand this can avoid unnecessary re-renders (arbitrarily many) when the other alpha value changes.